### PR TITLE
Add a test for wAsync behavior check

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fixes: #1224

File: test/lib/Streamly/Test/Prelude/Common.hs [hunk 1]
```diff
@@ -290,6 +291,18 @@ constructWithFromIndicesM op len =
         streamEffect <- run $ readIORef mvl
         listEquals (==) streamEffect list
 
+constructfromWAsync ::
+    S.WAsyncT IO Int -> S.WAsyncT IO Int-> [Int] -> Property
+constructfromWAsync s1 s2 res =
```
Comment (adithyaov) [inline test/lib/Streamly/Test/Prelude/Common.hs:296]: Can you also check the asyncness of the code?
An `IORef` to capture the effects would work.
```
gen i = threadDelay i >> addToIORefStuff >> return i
```
Then you can use `fromListM` and check the IORef for the expected order of effects.

File: test/Streamly/Test/Prelude/Async.hs [hunk 2]
```diff
@@ -56,6 +56,14 @@ main = hspec
...
         asyncOps $ prop "asyncly (.:)" . constructWithCons (S..:)
         asyncOps $ prop "asyncly (|:)" . constructWithConsM (S.|:) sort
+        prop "asyncSingleThreaded" $
+            constructfromAsyncSingleThread
+            (S.fromList [1,2,3,4,5])
+            (S.fromList [6,7,8,9,10])
+            [1,2,3,4,5,6,7,8,9,10]
+
+        prop "asyncMultiThreaded" $
+            constructfromAsyncNThread 4 2 [2,4]
```
Comment (harendra-kumar) [inline test/Streamly/Test/Prelude/Async.hs:66]: This is already being at line 127 below using `parallelCheck`. We can remove it.

File: test/lib/Streamly/Test/Prelude/Common.hs [hunk 3]
```diff
@@ -290,6 +293,41 @@ constructWithFromIndicesM op len =
...
+    withMaxSuccess maxTestCount $
+    monadicIO $ do
+        x <-  run
+            $ S.toList
+            $ S.fromWAsync
+            $ S.maxThreads 1
+            $ s1 `S.wAsync` s2
+        equals (==) x res
```
Comment (harendra-kumar) [inline test/lib/Streamly/Test/Prelude/Common.hs:306]: We can move this code to WAsync test module itself as this is specific to wAsync and not common.

File: test/lib/Streamly/Test/Prelude/Common.hs [hunk 4]
```diff
@@ -290,6 +293,41 @@ constructWithFromIndicesM op len =
...
+    S.AsyncT IO Int -> S.AsyncT IO Int-> [Int] -> Property
+constructfromAsyncSingleThread s1 s2 res =
+    withMaxSuccess maxTestCount $
+    monadicIO $ do
+        x <-  run
+            $ S.toList
+            $ S.fromAsync
+            $ S.maxThreads 1
+            $ s1 `S.async` s2
+        equals (==) x res
```
Comment (harendra-kumar) [inline test/lib/Streamly/Test/Prelude/Common.hs:318]: We can move this code to Async test module itself as this is specific to Async and not common.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Pull request merged**